scipy/stats/_morestats.py

```diff
@@ -1861,19 +1861,19 @@ def shapiro(x):
       and false negatives (failure to reject a false null hypothesis).
 
     """
-    x = np.ravel(x)
+    x = np.ravel(x).astype(np.float64)
 
     N = len(x)
     if N < 3:
         raise ValueError("Data must be at least length 3.")
 
     x = x - np.median(x)
 
-    a = zeros(N, 'f')
+    a = zeros(N//2, dtype=np.float64)
     init = 0
 
     y = sort(x)
```

Comment:
Fine to modify `y` in place since it's a copy.
```suggestion
    y = sort(x)
    y -= x[N//2]  # subtract the median (or a nearby value); see gh-15777
```

Reply:
Done.